Update k8s dependencies to 1.33.1 and bump go to 1.24

fasmat opened the pull request:
## What problem does this PR solve?

This PR updates the k8s dependencies of chaos-mesh to the newest released version and updates the code to not use now deprecated and removed functionality from the k8s libraries.

## What's changed and how it works?

I used the `hack/update-kubernetes-library.sh` script to update the kubernetes libraries to version `v1.32.2` and fixed all issues that popped up in the codebase because of it, primarily related to the removal of the two interfaces `webhook.Validator` and `webhook.Defaulter` which have been replaced by `webhook.CustomValidator` and `webhook.CustomDefaulter` respectivally.

Additionally `framework.ExpectError` and similar assertion functions have been removed from the kubernetes libraries. I replaced them with the underlying `gomega` code they would have called to preserve the tests that used them.

## Related changes

- [ ] This change also requires further updates to the [website](https://github.com/chaos-mesh/website) (e.g. docs)
- [ ] This change also requires further updates to the `UI interface`

## Cherry-pick to release branches (optional)

> This PR should be cherry-picked to the following release branches:

- [ ] release-2.7
- [ ] release-2.6

I am not sure about this, but I would like to suggest to also tag `api/v2.7.2` when releasing `v2.7.2` (and future versions) since right now the `api` module isn't versioned and users of it have to make sure to `go get` the right commit based on which chaos-mesh version they are using.

## Checklist

### CHANGELOG

> Must include at least one of them.

- [x] I have updated the `CHANGELOG.md`
- [ ] I have labeled this PR with "no-need-update-changelog"

### Tests

> Must include at least one of them.

- [x] Unit test
- [x] E2E test
- [ ] Manual test

Tests have been updated where deprecated code was used.

### Side effects

- [ ] **Breaking backward compatibility**

## DCO

If you find the DCO check fails, please run commands like below to fix it:

> [!TIP]
> Depends on actual situations, for example, if the failed commit isn't the most recent
> one, you can use `git rebase -i HEAD~n` to re-signoff the commit.

```shell
git commit --amend --signoff
git push --force
```

g1eny0ung commented:
I'm not sure why, I simply upgraded the k8s version of the e2e test, but the test still fails. I need to troubleshoot a bit more.

fasmat commented:
Thank you @g1eny0ung! To my understanding `mdnschaos.kb.io` is deployed with the `install.sh` script which I didn't modify by hand but tried updating with `hack/update_install_script.sh` (but it doesn't seem to do anything). The script still references an older version of k8s so maybe that's the issue?

fasmat commented:
> Hi @fasmat, I checked the changed webhook code and made some fixes. The reason for `failed to call webhook: the server could not find the requested resource` is we need to mount the validators on the webhook manager, you can see the changes in [84a46d2](https://github.com/chaos-mesh/chaos-mesh/pull/4644/commits/84a46d2f24a0d46185b25bdd1c4dd36c843d95be).

Good find! This wasn't even on my radar when I stumbled upon the problem.

> Also, I removed the snapshot file changes in `ui`, this is controlled by libraries, so you don't need to manually remove spaces in it. One suggestion I have is that if you find some formatting issues, it's best to go ahead and open a separate PR so that less attention is paid to minor changes.

This must have been accidental when I fixed spacing in other files; I will try to limit those changes to a smaller scope in future changes 🙂

STRRL commented:
I am going to try to fix the webhook part :)

STRRL commented:
it seems the webhook are not work as expected:

```
2025-03-29T23:01:42.677Z	INFO	setup	chaos-controller-manager/main.go:174	Setting up webhook server
2025-03-29T23:01:42.677Z	INFO	controller-runtime.webhook	webhook/server.go:183	Registering webhook	{"path": "/validate-auth"}
2025-03-29T23:01:42.677Z	INFO	setup	chaos-controller-manager/main.go:206	Starting manager
2025-03-29T23:01:42.677Z	INFO	controller-runtime.metrics	server/server.go:208	Starting metrics server
2025-03-29T23:01:42.677Z	INFO	setup	chaos-controller-manager/main.go:193	setup ctrlserver	{"addr": ":10082"}
2025-03-29T23:01:42.677Z	INFO	controller-runtime.metrics	server/server.go:247	Serving metrics server	{"bindAddress": "0.0.0.0:10080", "secure": false}
2025-03-29T23:01:42.677Z	INFO	controller-runtime.webhook	webhook/server.go:191	Starting webhook server
I0329 23:01:42.677752       1 leaderelection.go:257] attempting to acquire leader lease chaos-mesh-debug/chaos-mesh...
2025-03-29T23:01:42.678Z	INFO	controller-runtime.certwatcher	certwatcher/certwatcher.go:211	Updated current TLS certificate
2025-03-29T23:01:42.678Z	INFO	controller-runtime.webhook	webhook/server.go:242	Serving webhook server	{"host": "0.0.0.0", "port": 10250}
2025-03-29T23:01:42.678Z	INFO	controller-runtime.certwatcher	certwatcher/certwatcher.go:133	Starting certificate poll+watcher	{"interval": "10s"}
```

```
k -n chaos-mesh-debug port-forward chaos-controller-manager-6b97f7c667-n9w4t 8443:10250
Forwarding from 127.0.0.1:8443 -> 10250
Forwarding from [::1]:8443 -> 10250
```

```
~/playground/GitHub/chaos-mesh/chaos-mesh fix/fix-webhooks-with-pr-4644 ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── strrl@eium3 17:01:52
❯ curl --insecure -L https://localhost:8443/validate-auth
{"response":{"uid":"","allowed":false,"status":{"metadata":{},"message":"contentType=, expected application/json","code":400}}}

~/playground/GitHub/chaos-mesh/chaos-mesh fix/fix-webhooks-with-pr-4644 ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── strrl@eium3 17:02:58
❯ curl --insecure -L https://localhost:8443/validate-chaos-mesh-org-v1alpha1-podchaos
404 page not found

~/playground/GitHub/chaos-mesh/chaos-mesh fix/fix-webhooks-with-pr-4644 ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── strrl@eium3 17:03:07
❯ curl --insecure -L https://localhost:8443/mutate-chaos-mesh-org-v1alpha1-podchaos
404 page not found

```


As log says, our customized webhook `/validate-auth` still works, but general validation / mutation get 404

STRRL commented:
OK I get that,

https://github.com/kubernetes-sigs/controller-runtime/pull/2639
https://github.com/kubernetes-sigs/controller-runtime/pull/2877

We have to migrate to customzied validator / customzied mutator, becuase the origin interface have been removed....

I will follow @g1eny0ung's work

STRRL commented:
> I don't know how to fix the markdown links lint issue (the links seem to be valid) and I am not sure what is wrong with the failing integration test 🤔

I think I should update the markdown link linter config, add more pattern to ignore some URLs;

I am working on fixing integration tests.

STRRL commented:
bugbot run

STRRL reviewed (approved):
LGTM! 

Huge thanks @fasmat @g1eny0ung
